proposals/0000-specify-dump-json.md

```diff
@@ -0,0 +1,155 @@
+
+What follows is an explanation of the diagnostic message's schema. To demonstrate the appropriateness of the proposed schema, the internal diagnostic representation is produced below. Comments are removed for brevity.
+```haskell
+-- compiler/GHC/Types/Error.hs
+data MsgEnvelope e = MsgEnvelope
+   { errMsgSpan        :: SrcSpan
+   , errMsgContext     :: NamePprCtx
+   , errMsgDiagnostic  :: e
+   , errMsgSeverity    :: Severity
+   } deriving (Functor, Foldable, Traversable) 
+-- compiler/GHC/Types/Error.hs
+class (HasDefaultDiagnosticOpts (DiagnosticOpts a)) => Diagnostic a where
+  type DiagnosticOpts a
+  diagnosticMessage :: DiagnosticOpts a -> a -> DecoratedSDoc
+  diagnosticReason  :: a -> DiagnosticReason
+  -- | Extract any hints a user might use to repair their
+  -- code to avoid this diagnostic.
+  diagnosticHints   :: a -> [GhcHint]
+  diagnosticCode    :: a -> Maybe DiagnosticCode
+```
+All of the above is explained in the [wiki](https://gitlab.haskell.org/ghc/ghc/-/wikis/Errors-as-(structured)-values).
+
+The output of the JSON dump would consist of a list of errors, as well as a top level version. Each error contains the following fields:
+ - `"span"` corresponding to the above `SrcSpan`
+   - required by the schema
+   - This contains sub-fields, specified in the schema, which contains line numbers, etc.
+ - `"severity"` corresponding to the above `Severity`
+   - required by the schema
+ - `"code"` corresponding to the change in this [proposal](https://github.com/haskellfoundation/tech-proposals/blob/main/proposals/accepted/024-error-messages.md) and the above `DiagnosticCode` in typeclass `Diagnostic`
+   -  this is required by the schema but is allowed to have a value of `null`. The hope is that once all diagnostics officially have error codes, this can be required and nonnullable, but that will happen once `diagnosticCode` itself doesn't return a `DiagnosticCode` wrapped in a `Maybe`
+ - `"hints"` corresponding to the above `[GhcHint]`
+   - The schema itself doesn't validate at all what these may be, only that they are themselves objects. In my view, the best way to handle them is to output a JSON-ified version of the type `GhcHint`. Any attempt to specify them more than that will result in spiraling complexity and a required update every time the `GhcHint` type changes. Another alternative is to just remove them from the output and only present suggestions as they are rendered in the `message`.
```

Comment:
Doesn't this rather reduce the usefulness? One example usage of this that people suggested was an "auto-fixer" for issues. Presumably that would run off the hints, and if they're not specified then people are back to ad-hoc string parsing.

Reply:
That is a good point. In that case, what do you think about presenting a JSONified version of GHC's internal 'GhcHint' type?

Reply:
I'm not going to be an actual consumer of this so I'm not sure really :shrug: I think part of the issue is that then you presumably will have to either get GHC devs to stabilise the hint type or change the schema frequently.